argocd/tests/test_argocd.py

```diff
@@ -0,0 +1,85 @@
+    mock_http_response(file_path=get_fixture_path('app_controller_metrics.txt'))
+    check = ArgocdCheck('argocd', {}, [common.MOCKED_APP_CONTROLLER_INSTANCE])
+    dd_run_check(check)
+
+    for metric in common.APP_CONTROLLER_METRICS:
+        if metric in common.NOT_EXPOSED_METRICS:
+            aggregator.assert_metric(metric, at_least=0)
+        elif metric == 'argocd.app_controller.go.memstats.alloc_bytes':
+            aggregator.assert_metric(metric, metric_type=aggregator.GAUGE)
+        else:
+            aggregator.assert_metric(metric)
+
+    aggregator.assert_all_metrics_covered()
+
+
+def test_api_server(dd_run_check, aggregator, mock_http_response):
+    mock_http_response(file_path=get_fixture_path('api_server_metrics.txt'))
+    check = ArgocdCheck('argocd', {}, [common.MOCKED_API_SERVER_INSTANCE])
+    dd_run_check(check)
+
+    for metric in common.API_SERVER_METRICS:
+        if metric in common.NOT_EXPOSED_METRICS:
+            aggregator.assert_metric(metric, at_least=0)
+        elif metric == 'argocd.api_server.go.memstats.alloc_bytes':
+            aggregator.assert_metric(metric, metric_type=aggregator.GAUGE)
+        else:
+            aggregator.assert_metric(metric)
+
+    aggregator.assert_all_metrics_covered()
+
+
+def test_repo_server(dd_run_check, aggregator, mock_http_response):
+    mock_http_response(file_path=get_fixture_path('repo_server_metrics.txt'))
+    check = ArgocdCheck('argocd', {}, [common.MOCKED_REPO_SERVER_INSTANCE])
+    dd_run_check(check)
+
+    for metric in common.REPO_SERVER_METRICS:
+        if metric in common.NOT_EXPOSED_METRICS:
+            aggregator.assert_metric(metric, at_least=0)
+        elif metric == 'argocd.repo_server.go.memstats.alloc_bytes':
+            aggregator.assert_metric(metric, metric_type=aggregator.GAUGE)
+        else:
+            aggregator.assert_metric(metric)
+
+    aggregator.assert_all_metrics_covered()
+
+
+def test_empty_instance(dd_run_check):
+    try:
+        check = ArgocdCheck('argocd', {}, [{}])
+        dd_run_check(check)
+    except Exception as e:
+        assert "Must specify at least one of the following:`app_controller_endpoint`, `repo_server_endpoint` or" in str(
+            e
+        )
+
+
+def test_app_controller_service_check(dd_run_check, aggregator, mock_http_response):
+    # Test for transformer. The prometheus source submits a 1 or 0. 1 being OK and 0 being CRITICAL.
+    mock_http_response(file_path=get_fixture_path('app_controller_metrics.txt'))
+    check = ArgocdCheck('argocd', {}, [common.MOCKED_APP_CONTROLLER_INSTANCE])
+    dd_run_check(check)
+
+    aggregator.assert_service_check(
+        'argocd.app_controller.cluster.connection.status',
+        ServiceCheck.OK,
+        tags=['endpoint:app_controller:8082', 'name:bar'],
+    )
+    aggregator.assert_service_check(
+        'argocd.app_controller.cluster.connection.status',
+        ServiceCheck.CRITICAL,
+        tags=['endpoint:app_controller:8082', 'name:foo'],
```

Comment:
For my understanding, why is the `name:foo` tag different here compared to the `OK` check? This may benefit from `pytest.mark.parametrize` too btw ;)

Reply:
If I understand you correctly. Then the service checks here are different because the value the openmetrics payload is providing:

```
# HELP argocd_cluster_connection_status manually added
# TYPE argocd_cluster_connection_status gauge
argocd_cluster_connection_status{name="foo"} 0
argocd_cluster_connection_status{name="bar"} 1
```

The transformer here takes 0 as a `CRITICAL` and 1 as `OK`. So this test makes sure that both cases are covered.

Reply:
I was thinking of ways to parameterize this, but can't find a good way to do so. I think both service checks (4 now that I added a few more test cases for unknown), are generated from 1 scrape, I think it would make more sense to assert them all at once rather than assert them individually. Let me know what you think.